src/main/java/bio/terra/service/snapshot/SnapshotService.java

```diff
@@ -373,6 +374,30 @@ public EnumerateSnapshotModel enumerateSnapshots(
         .errors(errors);
   }
 
+  /**
+   * @param userReq authenticated user
+   * @return accessible snapshot IDs mapped to the roles which confer access
+   */
+  public SnapshotIdsAndRolesModel getSnapshotIdsAndRoles(AuthenticatedUserRequest userReq) {
+    List<ErrorModel> errors = new ArrayList<>();
+    Map<UUID, Set<IamRole>> authorizedSnapshots = listAuthorizedSnapshots(userReq, errors);
+
+    // We could have multiple TDRs talking to the same Sam (as for dev environments),
+    // so should only return authorized snapshot UUIDs also present in TDR.
+    Set<UUID> tdrSnapshotUuids = new HashSet<>(snapshotDao.getSnapshotIds());
+
+    Map<String, List<String>> roleMap =
+        authorizedSnapshots.keySet().stream()
+            .filter(tdrSnapshotUuids::contains)
```

Comment:
I loaded `tdrSnapshotUuids` into a `HashSet` so that we could call `contains` on it efficiently -- `O(1)`.

In production, the caller's authorized snapshots will be a smaller set than all of the `tdrSnapshotUuids` since TDR prod is the only TDR to load snapshots into Sam prod.  So I elected to iterate over those authorized snapshots.